File: src/main/java/io/lsdconsulting/lsd/distributed/interceptor/captor/http/RequestCaptor.java
package io.lsdconsulting.lsd.distributed.interceptor.captor.http;

import feign.Request;
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction;
import io.lsdconsulting.lsd.distributed.interceptor.captor.convert.TypeConverter;
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.PathDeriver;
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.SourceTargetDeriver;
import io.lsdconsulting.lsd.distributed.interceptor.captor.trace.TraceIdRetriever;
import io.lsdconsulting.lsd.distributed.interceptor.persistance.RepositoryService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpRequest;

import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.Collection;
import java.util.Map;

import static io.lsdconsulting.lsd.distributed.access.model.InteractionType.REQUEST;
import static java.util.Collections.emptyMap;

@Slf4j
@RequiredArgsConstructor
public class RequestCaptor {

    private final RepositoryService repositoryService;
    private final SourceTargetDeriver sourceTargetDeriver;
    private final PathDeriver pathDeriver;
    private final TraceIdRetriever traceIdRetriever;
    private final HttpHeaderRetriever httpHeaderRetriever;
    private final String profile;

    public InterceptedInteraction captureRequestInteraction(final Request request) {
        final var headers = httpHeaderRetriever.retrieve(request);
        final String body = TypeConverter.convert(request.body());
        final String path = pathDeriver.derivePathFrom(request.url());
        final String traceId = traceIdRetriever.getTraceId(headers);
        final String target = sourceTargetDeriver.deriveTarget(headers, path);
        final String serviceName = sourceTargetDeriver.deriveServiceName(headers);
        final InterceptedInteraction interceptedInteraction = buildInterceptedInteraction(headers, body, path, traceId, target, serviceName, request.httpMethod().name());
        repositoryService.enqueue(interceptedInteraction);
        return interceptedInteraction;
    }

    public InterceptedInteraction captureRequestInteraction(final HttpRequest request, final String body) {
        final var headers = httpHeaderRetriever.retrieve(request);
        final String path = pathDeriver.derivePathFrom(request);
        final String traceId = traceIdRetriever.getTraceId(headers);
        final String target = sourceTargetDeriver.deriveTarget(headers, path);
        final String serviceName = sourceTargetDeriver.deriveServiceName(headers);
        final InterceptedInteraction interceptedInteraction = buildInterceptedInteraction(headers, body, path, traceId, target, serviceName, request.getMethodValue());
        repositoryService.enqueue(interceptedInteraction);
        return interceptedInteraction;
    }

    private InterceptedInteraction buildInterceptedInteraction(Map<String, Collection<String>> headers, String body, String path, String traceId, String target, String serviceName, String httpMethod) {
        return new InterceptedInteraction(
                traceId,
                body,
                headers,
                emptyMap(),
                serviceName,
                target,
                path,
                null,
                httpMethod,
                REQUEST,
                profile,
                0L,
                ZonedDateTime.now(ZoneId.of("UTC")));
    }
}

File: src/main/kotlin/io/lsdconsulting/lsd/distributed/interceptor/captor/http/RequestCaptor.kt
package io.lsdconsulting.lsd.distributed.interceptor.captor.http

import feign.Request
import io.lsdconsulting.lsd.distributed.access.model.InteractionType
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction
import io.lsdconsulting.lsd.distributed.interceptor.captor.convert.convert
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.PathDeriver
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.derive.SourceTargetDeriver
import io.lsdconsulting.lsd.distributed.interceptor.captor.trace.TraceIdRetriever
import io.lsdconsulting.lsd.distributed.interceptor.persistance.RepositoryService
import org.springframework.http.HttpRequest
import java.time.ZoneId
import java.time.ZonedDateTime

class RequestCaptor(
    private val repositoryService: RepositoryService,
    private val sourceTargetDeriver: SourceTargetDeriver,
    private val pathDeriver: PathDeriver,
    private val traceIdRetriever: TraceIdRetriever,
    private val httpHeaderRetriever: HttpHeaderRetriever,
    private val profile: String,
){
    fun captureRequestInteraction(request: Request): InterceptedInteraction {
        val headers = httpHeaderRetriever.retrieve(request)
        val body = request.body()?.convert()
        val path = pathDeriver.derivePathFrom(request.url())
        val traceId = traceIdRetriever.getTraceId(headers)
        val target = sourceTargetDeriver.deriveTarget(headers, path)
        val serviceName = sourceTargetDeriver.deriveServiceName(headers)
        val interceptedInteraction =
            buildInterceptedInteraction(headers, body, path, traceId, target, serviceName, request.httpMethod().name)
        repositoryService.enqueue(interceptedInteraction)
        return interceptedInteraction
    }

    fun captureRequestInteraction(request: HttpRequest, body: String?): InterceptedInteraction {
        val headers = httpHeaderRetriever.retrieve(request)
        val path = pathDeriver.derivePathFrom(request)
        val traceId = traceIdRetriever.getTraceId(headers)
        val target = sourceTargetDeriver.deriveTarget(headers, path)
        val serviceName = sourceTargetDeriver.deriveServiceName(headers)
        val interceptedInteraction =
            buildInterceptedInteraction(headers, body, path, traceId, target, serviceName, request.methodValue)
        repositoryService.enqueue(interceptedInteraction)
        return interceptedInteraction
    }

    private fun buildInterceptedInteraction(
        headers: Map<String, Collection<String>>,
        body: String?,
        path: String,
        traceId: String,
        target: String,
        serviceName: String,
        httpMethod: String?
    ) = InterceptedInteraction(
        traceId = traceId,
        body = body,
        requestHeaders = headers,
        responseHeaders = emptyMap(),
        serviceName = serviceName,
        target = target,
        path = path,
        httpStatus = null,
        httpMethod = httpMethod, interactionType = InteractionType.REQUEST,
        profile = profile,
        elapsedTime = 0L,
        createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
    )
}
